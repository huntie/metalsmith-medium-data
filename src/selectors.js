import { pick } from 'lodash';

/**
 * Select a transformed subset of post data from the given response payload.
 *
 * @param {Object} payload
 *
 * @return {Array}
 */
export const getPosts = payload =>
  Object.values(payload.references.Post)
    .map(post => ({
      ...pick(post, [
        'canonicalUrl',
        'coverless',
        'createdAt',
        'creatorId',
        'detectedLanguage',
        'displayAuthor',
        'firstPublishedAt',
        'inResponseToPostId',
        'inResponseToRemovedAt',
        'isNsfw',
        'isProxyPost',
        'isSeries',
        'latestPublishedAt',
        'latestPublishedVersion',
        'license',
        'slug',
        'title',
        'type',
        'versionId'
      ]),
      subtitle: post.previewContent.subtitle,
      url: `https://medium.com/@${payload.user.username}/${post.uniqueSlug}`,
      ...pick(post.virtuals, [
        'previewImage',
        'readingTime',
        'recommends',
        'responsesCreatedCount',
        'totalClapCount'
      ]),
      tags: post.virtuals.tags.map(tag => pick(tag, ['name', 'slug']))
    }));
