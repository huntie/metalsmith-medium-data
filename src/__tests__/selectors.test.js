import { last } from 'lodash';
import snapshotDiff from 'snapshot-diff';
import { getPosts } from '../selectors';

describe('selectors', () => {
  describe('getPosts', () => {
    test('should select a transformed subset of post data', () => {
      const { payload } = require('./sample-response.json');

      expect(
        snapshotDiff(
          last(Object.values(payload.references.Post)),
          last(getPosts(payload))
        )
      ).toMatchSnapshot();
    });
  });
});
