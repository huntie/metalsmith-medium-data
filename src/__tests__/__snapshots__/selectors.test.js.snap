// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors getPosts should select a transformed subset of post data 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
-   \\"acceptedAt\\": 0,
-   \\"allowResponses\\": true,
-   \\"approvedHomeCollectionId\\": \\"\\",
-   \\"audioVersionDurationSec\\": 0,
    \\"canonicalUrl\\": \\"\\",
-   \\"content\\": Object {
-     \\"postDisplay\\": Object {
-       \\"coverless\\": true,
-     },
-     \\"subtitle\\": \\"UUIDs are awesome — they provide a non-sequential, comprehensively unique identifier that can be generated separately from the source…\\",
-   },
    \\"coverless\\": true,
    \\"createdAt\\": 1483703513624,
    \\"creatorId\\": \\"e54e8fe3cbbe\\",
-   \\"curationEligibleAt\\": 0,
-   \\"deletedAt\\": 0,
    \\"detectedLanguage\\": \\"en\\",
    \\"displayAuthor\\": \\"\\",
-   \\"editorialPreviewDek\\": \\"\\",
-   \\"editorialPreviewTitle\\": \\"\\",
-   \\"experimentalCss\\": \\"\\",
-   \\"featureLockRequestAcceptedAt\\": 0,
    \\"firstPublishedAt\\": 1483733066220,
-   \\"hasUnpublishedEdits\\": true,
-   \\"hightowerMinimumGuaranteeEndsAt\\": 0,
-   \\"hightowerMinimumGuaranteeStartsAt\\": 0,
-   \\"homeCollectionId\\": \\"\\",
-   \\"id\\": \\"c30e66bf35f9\\",
-   \\"importedPublishedAt\\": 0,
-   \\"importedUrl\\": \\"\\",
-   \\"inResponseToMediaResourceId\\": \\"\\",
    \\"inResponseToPostId\\": \\"\\",
    \\"inResponseToRemovedAt\\": 0,
-   \\"isApprovedTranslation\\": false,
-   \\"isBlockedFromHightower\\": false,
-   \\"isEligibleForRevenue\\": false,
    \\"isNsfw\\": false,
    \\"isProxyPost\\": false,
    \\"isSeries\\": false,
-   \\"isSubscriptionLocked\\": false,
-   \\"isTitleSynthesized\\": true,
    \\"latestPublishedAt\\": 1514394793685,
    \\"latestPublishedVersion\\": \\"677830f49db1\\",
-   \\"latestRev\\": 962,
-   \\"latestVersion\\": \\"677830f49db1\\",
-   \\"layerCake\\": 0,
    \\"license\\": 0,
-   \\"lockedPostSource\\": 0,
-   \\"mediumUrl\\": \\"\\",
-   \\"migrationId\\": \\"\\",
-   \\"mongerRequestType\\": 1,
-   \\"newsletterId\\": \\"\\",
-   \\"notifyFacebook\\": false,
-   \\"notifyFollowers\\": true,
-   \\"notifyTwitter\\": true,
-   \\"previewContent\\": Object {
-     \\"bodyModel\\": Object {
-       \\"paragraphs\\": Array [
-         Object {
-           \\"alignment\\": 1,
-           \\"markups\\": Array [],
-           \\"name\\": \\"c932\\",
-           \\"text\\": \\"Representing a UUID as a Base62 Hash ID for short, pretty URLs\\",
-           \\"type\\": 3,
-         },
-         Object {
-           \\"alignment\\": 1,
-           \\"markups\\": Array [],
-           \\"name\\": \\"8619\\",
-           \\"text\\": \\"UUIDs are awesome — they provide a non-sequential, comprehensively unique identifier that can be generated separately from the source database. However, if you need to include one in a URL they can appear somewhat…\\",
-           \\"type\\": 1,
-         },
-       ],
-       \\"sections\\": Array [
-         Object {
-           \\"startIndex\\": 0,
-         },
-       ],
-     },
-     \\"isFullContent\\": false,
-     \\"subtitle\\": \\"UUIDs are awesome — they provide a non-sequential, comprehensively unique identifier that can be generated separately from the source…\\",
-   },
-   \\"proxyPostFaviconUrl\\": \\"\\",
-   \\"proxyPostProviderName\\": \\"\\",
-   \\"responseHiddenOnParentPostAt\\": 0,
-   \\"sequenceId\\": \\"\\",
-   \\"seriesLastAppendedAt\\": 0,
-   \\"slug\\": \\"representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls\\",
-   \\"socialDek\\": \\"\\",
-   \\"socialTitle\\": \\"\\",
-   \\"title\\": \\"Representing a UUID as a Base62 Hash ID for short, pretty URLs\\",
-   \\"translationSourceCreatorId\\": \\"\\",
-   \\"translationSourcePostId\\": \\"\\",
-   \\"type\\": \\"Post\\",
-   \\"uniqueSlug\\": \\"representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls-c30e66bf35f9\\",
-   \\"updatedAt\\": 1529307247534,
-   \\"versionId\\": \\"677830f49db1\\",
-   \\"virtuals\\": Object {
-     \\"allowNotes\\": true,
-     \\"imageCount\\": 0,
-     \\"isBookmarked\\": false,
-     \\"isLockedPreviewOnly\\": false,
-     \\"links\\": Object {
-       \\"entries\\": Array [
-         Object {
-           \\"alts\\": Array [],
-           \\"httpStatus\\": 200,
-           \\"url\\": \\"http://hashids.org/\\",
-         },
-         Object {
-           \\"alts\\": Array [],
-           \\"httpStatus\\": 200,
-           \\"url\\": \\"https://eager.io/blog/how-long-does-an-id-need-to-be/\\",
-         },
-       ],
-       \\"generatedAt\\": 1514394794201,
-       \\"version\\": \\"0.3\\",
-     },
-     \\"metaDescription\\": \\"\\",
-     \\"noIndex\\": false,
    \\"previewImage\\": Object {
      \\"backgroundSize\\": \\"\\",
      \\"filter\\": \\"\\",
      \\"height\\": 0,
      \\"imageId\\": \\"\\",
      \\"originalHeight\\": 0,
      \\"originalWidth\\": 0,
      \\"strategy\\": \\"resample\\",
      \\"width\\": 0,
    },
-     \\"readingList\\": 0,
    \\"readingTime\\": 2.2037735849056603,
    \\"recommends\\": 39,
    \\"responsesCreatedCount\\": 0,
-     \\"sectionCount\\": 1,
-     \\"socialRecommendsCount\\": 0,
+   \\"slug\\": \\"representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls\\",
    \\"subtitle\\": \\"UUIDs are awesome — they provide a non-sequential, comprehensively unique identifier that can be generated separately from the source…\\",
    \\"tags\\": Array [
      Object {
-         \\"metadata\\": Object {
-           \\"coverImage\\": Object {
-             \\"id\\": \\"1*B38rwiQOcVjtwQy988CfJA.png\\",
-             \\"isFeatured\\": true,
-             \\"originalHeight\\": 312,
-             \\"originalWidth\\": 820,
-           },
-           \\"postCount\\": 99880,
-         },
        \\"name\\": \\"Programming\\",
-         \\"postCount\\": 99880,
        \\"slug\\": \\"programming\\",
-         \\"type\\": \\"Tag\\",
      },
      Object {
-         \\"metadata\\": Object {
-           \\"coverImage\\": Object {
-             \\"id\\": \\"1*_eFHsjMYw7uNLzSHOF7_AA.png\\",
-             \\"isFeatured\\": true,
-             \\"originalHeight\\": 720,
-             \\"originalWidth\\": 1280,
-           },
-           \\"postCount\\": 11876,
-         },
        \\"name\\": \\"PHP\\",
-         \\"postCount\\": 11876,
        \\"slug\\": \\"php\\",
-         \\"type\\": \\"Tag\\",
      },
      Object {
-         \\"metadata\\": Object {
-           \\"coverImage\\": Object {
-             \\"id\\": \\"1*_eFHsjMYw7uNLzSHOF7_AA.png\\",
-             \\"isFeatured\\": true,
-             \\"originalHeight\\": 720,
-             \\"originalWidth\\": 1280,
-           },
-           \\"postCount\\": 6530,
-         },
        \\"name\\": \\"Laravel\\",
-         \\"postCount\\": 6530,
        \\"slug\\": \\"laravel\\",
-         \\"type\\": \\"Tag\\",
      },
      Object {
-         \\"metadata\\": Object {
-           \\"coverImage\\": Object {
-             \\"id\\": \\"1*FmlVUU9ozKJhEXoVtrVZsg.png\\",
-             \\"isFeatured\\": true,
-             \\"originalHeight\\": 600,
-             \\"originalWidth\\": 1800,
-           },
-           \\"postCount\\": 3,
-         },
        \\"name\\": \\"Hashids\\",
-         \\"postCount\\": 3,
        \\"slug\\": \\"hashids\\",
-         \\"type\\": \\"Tag\\",
      },
      Object {
-         \\"metadata\\": Object {
-           \\"postCount\\": 1,
-         },
        \\"name\\": \\"Uuids\\",
-         \\"postCount\\": 1,
        \\"slug\\": \\"uuid\\",
-         \\"type\\": \\"Tag\\",
      },
    ],
-     \\"topics\\": Array [],
+   \\"title\\": \\"Representing a UUID as a Base62 Hash ID for short, pretty URLs\\",
    \\"totalClapCount\\": 120,
-     \\"usersBySocialRecommends\\": Array [],
-     \\"wordCount\\": 584,
-   },
-   \\"visibility\\": 0,
-   \\"vote\\": false,
-   \\"webCanonicalUrl\\": \\"\\",
+   \\"type\\": \\"Post\\",
+   \\"url\\": \\"https://medium.com/@huntie/representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls-c30e66bf35f9\\",
+   \\"versionId\\": \\"677830f49db1\\",
  }"
`;
